Change top feed notice.

pratik-pc opened the pull request:
This fixes the wrong compose tooltip when compose is disabled. Disables compose reply button in non message list view.

This PR also changes top notice/narrow banner in:

## No search results
- Removes top of feed notice
- adds `Search all public channels` button with same href as top feed text link.

## Search results
- At the top of the message feed, when message feed notice is shown, removes `!` icon, background/border and changes text to 
> End of results from [your message history](https://github.com/help/search-for-messages#searching-shared-history).`
- Adds `Search all public channels` button aligned right

## Combined feed
At the top of the combined feed
- Adds notice text aligned left
- adds `View all messages in public channels` button to `#narrow/channels/public` aligned right

<details>
<summary>Commit description</summary>

Commit 1-7:

Refactors how the closed compose reply button tracks and updates its recipient context, label, and disabled state. Previously, the reply button label and compose permission state were derived through separate code paths. The label was updated from `recipient_information` passed to `get_recipient_label`(handling various other cases) by callers, while the disabled state was calculated with the reply target from the selected message or current narrow state only. This made compose disabled state to not work for inbox view and recent conversation view. Now, we store recipient metadata from `RecipientLabel` in reply button element wrapper and use it to compute disabled state.

Commit 8: We can now disable and update tooltip for compose button in recent view/inbox view using this metadata.

commit 9-10: Simple refactor into object.

Commit 11-14: Banner text html for some narrows.

Commit 16-18:

Redesign top of feed notice, empty narrow banner and make them mutually exclusive.

</details>


Fixes: zulip#33287
Fixes: zulip#32597

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<details>
<summary>Feed notice screenshots</summary>

|Before|After|
|-|-|
|`No search results`||
|<img width="929" height="254" alt="image" src="https://github.com/user-attachments/assets/27c5334f-c407-4010-9e96-3a284f03ebf4" />|<img width="906" height="404" alt="image" src="https://github.com/user-attachments/assets/1b474a02-6ca6-43dc-ba05-1c37ba58b398" />|
|`Top of search results`||
|<img width="940" height="233" alt="image" src="https://github.com/user-attachments/assets/f236e03c-1a59-4772-a639-835debbe85aa" />|<img width="936" height="436" alt="image" src="https://github.com/user-attachments/assets/b7965b66-bc28-4368-b19e-483572716de6" />|
||<img width="555" height="436" alt="image" src="https://github.com/user-attachments/assets/affdd767-d693-4159-81d8-b7a174789d63" />|
|`Top of combined feed`||
|<img width="940" height="184" alt="image" src="https://github.com/user-attachments/assets/ed780309-2507-463a-8899-7733a5970c4c" />|<img width="937" height="204" alt="image" src="https://github.com/user-attachments/assets/0393b5d2-ad05-422d-bfdf-25497c6b2a1b" />|
||<img width="495" height="254" alt="image" src="https://github.com/user-attachments/assets/9e787f91-864c-41d7-a258-f87b023517c3" />|


|Before|After|
|-|-|
|`has:`||
|<img width="929" height="254" alt="image" src="https://github.com/user-attachments/assets/7b646aa3-da49-46af-8152-48b588733c2f" />|<img width="938" height="300" alt="image" src="https://github.com/user-attachments/assets/c34bcf39-12eb-42e0-ab15-6e0b1ba16648" />|
|`sender:`||
|<img width="929" height="254" alt="image" src="https://github.com/user-attachments/assets/01af3040-caae-4f5c-ae44-d6b59f6cac1c" />|<img width="584" height="254" alt="image" src="https://github.com/user-attachments/assets/ab944fe9-f02d-4eab-ba81-217312148cad" />|
|`is:resolved`||
|<img width="929" height="254" alt="image" src="https://github.com/user-attachments/assets/50f4056c-bd8b-4df1-aba1-df85d68f0195" />|<img width="929" height="297" alt="image" src="https://github.com/user-attachments/assets/4b7637f7-72e7-4f7e-8d37-31e82a00c94d" />|
</details>

<details>
<summary>Compose reply button screenshot</summary>

|Before|After|
|-|-|
|<img width="934" height="626" alt="image" src="https://github.com/user-attachments/assets/6c78e76e-fb26-4a3a-a2c3-e374168345c2" />|<img width="934" height="626" alt="Screenshot from 2026-03-23 22-18-57" src="https://github.com/user-attachments/assets/7546008b-20c1-45e7-b163-5df8a264a680" />|
|<img width="754" height="399" alt="image" src="https://github.com/user-attachments/assets/86e7f1dd-ab1f-4409-a70d-c38bf773b166" />|<img width="754" height="399" alt="image" src="https://github.com/user-attachments/assets/5fab5330-0eb6-4fba-99e7-9fc63fc78d33" />|
|<img width="754" height="399" alt="image" src="https://github.com/user-attachments/assets/946ff3ac-899c-48e5-80b6-471f8f14dfcd" />|<img width="754" height="399" alt="image" src="https://github.com/user-attachments/assets/ffc6ae89-196e-478c-8120-34938bae0a7e" />|
|<img width="754" height="399" alt="image" src="https://github.com/user-attachments/assets/a962d0ef-d509-4e8b-831a-ef68a4e92db5" />|<img width="754" height="399" alt="image" src="https://github.com/user-attachments/assets/5ddaa639-ebec-4c76-8f55-c6ca026533ae" />|
</details>

<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

alya commented:
What's the relationship between this PR and #34788.

minhphung152 commented:
@alya I wasn't aware they were working on this issue. I was assigned this issue and have been working on it.

pratik-pc commented:
hm, I didn't see that PR. I saw it in priorities board and tried to clear it. I can do an initial review of your PR if it is up to date

pratik-pc committed:
narrow_banner: Add search shared history button to empty narrows.

When a narrow's message history may be incomplete, the top-of-feed
notice offers a way to search all public channels.

We move the same functionality to the empty-narrow banner, a
"Search all public channels" button that opens the corresponding
`channels:public` search, and links the banner title to the relevant
help center documentation.

We add a `show_action` field to show the button and
`should_show_search_shared_history_action()` decides whether to show
the button.

Search shared history button is shown only for valid narrows. Invalid
narrows are excluded because they would produce an equally invalid
`channels:public` search.

Also hide any existing top-of-feed notice when displaying an
empty-narrow banner so that the two UI elements are never shown
together.

Fixes part of #33287.

pratik-pc committed:
message_feed_notice: Redesign message search caution link top notice.

We remove the outline from the top notice and change the notice text.
Also, replaces the `searching all public channels` for the public
search link with a button `Expand search`.

Fixes part of zulip#33287.

pratik-pc committed:
combined_feed: Add top of feed notice to combined feed view.

Just like message search top notices, we also add a top notice in
combined feed view with text, no outline and a button
`View all messages in public channels`. Also, change
`may_have_incomplete_message_history` check to treat combined
feed narrow as incomplete.

Fixes: zulip#33287.

pratik-pc committed:
narrow_banner: Change banner text for empty `has` narrow.

We change the banner text from `There are no messages...` to
`No search results from your message history` when the narrow is empty.

pratik-pc commented:
@amanagr Are you the right person to review this?

alya commented:
Hm, yeah, I think we can try putting the button next to the text (not on the right), and letting it move to the next line rather than having the text take up 2 lines in narrow windows.

alya commented:
I'm not liking having both the `z` and "End of messages...`. Can we drop the `z` when this message is shown? We can still use the `z` as a loading indicator, but I assume we wouldn't show the "end of messages" notice while the view is loading?

alya commented:
While we're looking at this, let's change the `is:alerted` no results page to the default "No search results..." page. This is silly in that context:

<img width="963" height="298" alt="Screenshot 2026-02-16 at 22 06 36" src="https://github.com/user-attachments/assets/3509d015-056f-4c2b-a6c2-0c539b9fdc7e" />

pratik-pc committed:
filter: Refactor incomplete message history check for top end notice.

We render top end notice for some search narrows when we think not
all messages were searched. We determine this based on current narrow
filter. We refactor this incomplete messages in message list check
into a method on the Filter object.

pratik-pc committed:
narrow_banner: Change banner text for empty `is:alerted` narrow.

Changes the banner text from `There are no messages...` to
`No search results from your message history`.

pratik-pc committed:
narrow_banner: Change banner text for empty `is:resolved` narrow.

pratik-pc commented:
> Let's tweak the resolved topics one to "No topics in your history are marked as resolved.", as it's a bit confusing otherwise.
> 
> All the other screenshots look good to me!

Addressed this

@amanagr Ready for your review

amanagr commented on web/src/filter.ts:
```diff
@@ -1606,6 +1606,20 @@ export class Filter {
+    }
+
     contains_only_private_messages(): boolean {
         return (
```
Can this extraction be a prep commit?

amanagr commented on web/src/narrow_banner.ts:
```diff
@@ -552,16 +586,30 @@ export function pick_empty_narrow_banner(current_filter: Filter): NarrowBannerDa
     $(".empty_feed_notice_main").html(rendered_narrow_banner);
+    $(".all-messages-search-caution").hide();
+    $(".combined-feed-notice").hide();
+    $(".top-messages-logo").show();
```
Why do we need to do these here?

pratik-pc replied:
In a specific case, if the messages in the current narrow are removed after deletion or moving, we hide it since the empty banner is shown and already has its own button. I have added a comment for this

pratik-pc committed:
message_feed_top_notice: Use sorted_term_types method.

We replace accessing class variable with using method which returns
the value.

pratik-pc committed:
hash_util: Fix decode_dm_recipient_user_ids_from_narrow_url.

We should check for with/near operator on second term when second
term is not undefined since we sometimes have dm narrow url
in the format #narrow/dm-1,2

pratik-pc committed:
compose_closed_ui: Remove direct type from data-conversation-type.

`direct` type of data-conversation-type attribute for
`#new_conversation_button` is currently redundant. Earlier, it was
used to set the tooltip on #new_conversation_button separately for
DM and stream message views.

A separate DM button(#new_direct_message_button) was added for
starting dm conversations. Both the `non-specific` & `direct` types add
the `New channel message` tooltip, making the separation between
`non-specific` and `direct` type redundant.

pratik-pc committed:
inbox_ui: Update reply button recipient on inbox channel view.

The closed compose reply button logic to update recipient information
derives the stream_id from an inbox header's data-stream-id attribute.
This works in the general inbox view, but not in the "list of topics"
channel view, which does not have inbox headers. This results in
closed compose `.compose_reply_button` text to always show
`Compose message` instead of `Message #stream > topic` when a topic
is selected in the list of topics view.

We use the stream_id from the active channel filter via
`inbox_util.get_channel_id`.

pratik-pc committed:
compose_closed_ui: Compute reply button state from recipient context.

The closed compose reply button previously determined whether it
should be disabled by inspecting the current message list or narrow
state. This coupled the button state to the currently selected message
and required separate logic paths for stream and direct message
conversations.

Now that recipient metadata is stored on the reply button wrapper,
compute the button state directly from that recipient context. For
stream conversations, we use the stored stream_id to check posting
permissions; for direct messages, we use the stored recipient ids to
check whether the user can send messages to the conversation.

We change `should_disable_compose_reply_button_*` to check
permissions using the stream_id/user_ids derived from data attrs.

This decouples reply button state from the current message list and
narrow state, and allows the button to consistently reflect the
selected conversation across different views.

Fixes part of #32597.

pratik-pc committed:
compose_closed_ui: Disable compose reply in non conversation view.

Just like conversation views where we disable compose reply if the
user doesn't have appropriate permission to send message to the
message recipient, we also disable the compose reply in other views
(inbox ui, recent view, empty narrow).

We construct `recipient_information` from href(narrow urls), element
text, etc. and pass it to `update_reply_button` to reflect the
appropriate compose permissions. Also, updates compose tooltips
properly for disabled direct message compose button.

Fixes: zulip#32597.

pratik-pc commented on web/templates/message_feed_errors.hbs:
```diff
@@ -9,29 +9,43 @@
+        <p>
+            {{#tr}}
+            End of results from
+            <z-link>your message history</z-link>. Include messages in public channels
```
@alya what should the string be here? Since we’re also adding a new string/notice for the combined feed, should it be different as well?

alya replied:
Ah, great question. I think for guests, it should say "Include messages in your public channels" rather that "Include messages in public channels" -- that's what the button will do, right?

alya replied:
And similarly "View all messages in your public channels" as the button label for guests in the combined feed.

alya replied:
I guess this will all get a bit cleaner when we implement the `channels:all` filter and can use it here.

alya commented:
Looks good in light manual testing, so I think you can mark for integration review once strings for guests have been updated.

pratik-pc commented:
Removed merge conflicts

amanagr commented:
I pushed 4 commits for the changes I would liked for you to make in this pull request. Can you review them and squash into appropriate commits?

pratik-pc committed:
message_feed_top_notice: Inline empty narrow banner removal.

The history-limit notice and empty-narrow banner are mutually exclusive
top-of-feed states. Previously, we cleared the empty narrow banner by
calling into narrow_banner, but that dependency will cause an import
cycle in a subsequent commit.

Inline the banner clearing via
`$(".empty_feed_notice_main").empty()` instead.

pratik-pc commented:
@amanagr Squashed them into appropriate commits.

amanagr commented:
I pushed 3 more commits to which Claude identified after a deeper review. Can you review them?

pratik-pc commented on web/styles/zulip.css:
```diff
@@ -198,6 +198,9 @@ p.n-margin {
     display: none;
+    border-radius: 4px;
+    border: 1px solid hsl(192deg 19% 75% / 20%);
+    box-shadow: 0 0 2px hsl(192deg 19% 75% / 20%);
```
This is supposed to be a redesign that drops the outline, so this change isn't actually needed.

pratik-pc replied:
Fixed the commit message to mention the redesign.

pratik-pc commented:
Squashed the other 2 commits.

amanagr commented:
Great! LGTM! Requesting another review from @shubham-padia here.

shubham-padia commented on web/src/inbox_ui.ts:
```diff
@@ -1630,9 +1631,15 @@ function update_closed_compose_text($row: JQuery, is_header_row: boolean): void
+        } else {
+            const $stream = $row.parent(".inbox-topic-container").prev(".inbox-header");
+            stream_id = Number($stream.attr("data-stream-id"));
+        }
```
It's not clear from the commit message why this change is being made and what was the behaviour before. With a lot of changes in the PR, it makes it easy to review with more detail in each commit.

shubham-padia commented on web/src/compose_closed_ui.ts:
```diff
@@ -128,32 +132,35 @@ export function get_recipient_label(
+    } else if (user_ids_string !== undefined) {
+        disable = should_disable_compose_reply_button_for_direct_message(user_ids_string);
+    }
+
```
I'm unclear why we are doing all these changes in the commit `compose_closed_ui: Store recipient context as data attr on reply btn.`, can you explain more? Having more context in the commit would also be helpful

pratik-pc replied:
Split the computation changes into a separate commit.

shubham-padia reviewed (commented):
I'm having a tough time reviewing this PR, mainly due to not knowing why each commit is present where it is and what is it's purpose. Also, I see changes that don't necessarily match the commit description (like an extra refactor not mentioned in the commit message). I think it would be helpful for each commit to highlight the technical decisions being made and why they are being made. Some meta description of the commits in the PR description can also be helpful.

pratik-pc committed:
compose_closed_ui: Store recipient metadata in reply button label state.

The closed compose reply button previously rendered recipient labels
without preserving the underlying recipient information. As a result,
other reply button logic needed to re-derive the reply target from the
current message list or narrow state. For recent view and inbox view,
we can't derive much information about the selected table entry from
the narrow state alone.

Extend `RecipientLabel` to carry stream and direct message recipient
metadata, and store that information on the reply button wrapper via
`data-stream-id` and `data-user-ids-string` attributes when updating
the label.

This prepares for subsequent commit that will use the stored recipient
context directly when updating reply button state.

Fixes part of zulip#32597.

pratik-pc committed:
compose_closed_ui: Reuse reply button wrapper jQuery object.

Store the reply button wrapper element in a local variable rather
than repeating the same jQuery selector multiple times within
update_reply_button_state().

This simplifies the code and avoids repeated DOM lookups.

pratik-pc committed:
compose_closed_ui: Centralize reply button updates.

Introduce update_reply_button() to update the reply button recipient
context and state together.

This guarantees that update_reply_button_with_recipient_context()
runs before update_reply_button_state(), so that the latter can rely
on the recipient metadata stored on the button wrapper.

pratik-pc committed:
narrow_banner: Tighten `NarrowBannerData` type.

Change `NarrowBannerData` to include union of `title` and
`title_html`, but not both.

The empty-feed notice template renders only one of these fields, so
making them mutually exclusive better reflects the expected data shape.

We have a preference on showing `title_html` and as such we remove
the `title` field for empty mention banner.

pratik-pc commented:
Restructured the commits and updated commit messages. It should be easier to go through them now.

pratik-pc commented:
Resolved merge conflicts.

@shubham-padia Ready for review.

shubham-padia commented on web/src/inbox_ui.ts:
```diff
@@ -1630,9 +1631,19 @@ function update_closed_compose_text($row: JQuery, is_header_row: boolean): void
+        } else {
+            const $stream = $row.parent(".inbox-topic-container").prev(".inbox-header");
+            stream_id = Number($stream.attr("data-stream-id"));
+        }
```
get_focused_row_message uses inbox-topic-menu to get the channel id. Is that possible here as well?

pratik-pc replied:
Done.

shubham-padia commented on web/src/compose_closed_ui.ts:
```diff
@@ -232,12 +235,21 @@ function set_reply_button_label(label: string): void {
+    );
+    $compose_reply_button_wrapper.removeAttr("data-stream-id");
+    $compose_reply_button_wrapper.removeAttr("data-user-ids-string");
+    $compose_reply_button_wrapper.removeAttr("data-reply-button-type");
```
We should mention in a comment that we are clearing stale state

shubham-padia commented on web/src/compose_closed_ui.ts:
```diff
@@ -210,12 +212,13 @@ export function update_buttons(update_type?: string): void {
     update_new_conversation_button(update_type === "stream" ? "stream" : "non-specific");
+    // For interleaved views (e.g. combined feed, search, is:starred), the
+    // reply target is the selected message, which update_reply_button_with_recipient_context
+    // reads from the current message list when recipient_information is undefined.
```
I don't understand what this comment is trying to highlight.

pratik-pc replied:
Basically, we don't need to pass recipient_information here, since we derive it either from the selected message or from the current narrow state. For Inbox and Recent Conversations, we call the same function separately and pass recipient_information explicitly. Updated the comment for clarity.

shubham-padia commented on web/src/compose_closed_ui.ts:
```diff
@@ -248,6 +260,31 @@ export function update_recipient_text_for_reply_button(
         });
+        // Set data attributes so that update_reply_button_state can
+        // determine whether to disable the reply button without
+        // needing to re-examine the message list or narrow state.
```
A better comment would be something like:

```
We store the recipient state so that update_reply_button_state
can use that information without re-examining the message list
or narrow state. The state is reset/unset by xyz function(s).
```

pratik-pc replied:
Done

shubham-padia commented on web/tests/compose.test.cjs:
```diff
@@ -847,20 +847,16 @@ test_ui("DM policy disabled", ({override}) => {
-    override(narrow_state, "pm_ids_string", () => "31,33");
-    reply_disabled = compose_closed_ui.should_disable_compose_reply_button_for_direct_message();
+    reply_disabled =
+        compose_closed_ui.should_disable_compose_reply_button_for_direct_message("31,33");
```
I think the previous tests were wrong. For human user part should only have 33, right?

pratik-pc replied:
Yeah, fixed it.

shubham-padia commented on web/tests/message_view.test.cjs:
```diff
@@ -426,7 +429,7 @@ run_test("show_empty_narrow_message", ({mock_template, override, override_rewire
         empty_narrow_html(
             undefined,
-            `translated: No search results from <a href="/help/search-for-messages#searching-shared-history" target="_blank" rel="noopener noreferrer">your message history</a>.`,
+            `translated: No search results from <a href="/help/search-for-messages#search-shared-history" target="_blank" rel="noopener noreferrer">your message history</a>.`,
```
This change should belong in the previous commit

pratik-pc replied:
Done.

shubham-padia commented on web/src/narrow_banner.ts:
```diff
@@ -121,11 +142,12 @@ function empty_search_query_banner(current_filter: Filter): NarrowBannerData {
         return {
-            title: NO_SEARCH_RESULTS_TITLE,
+            ...no_search_results,
+            show_action: incomplete_history && !invalid_narrow,
```
I'm not a fan of this duplication of `incomplete_history && !invalid_narrow` everywhere. If show_action changes in the future, we have to amend this at all this individual pieces. 

Something like function 
```
show_search_shared_history_action(
    current_filter: Filter,
    invalid_narrow: boolean,
): boolean {
    return current_filter.may_have_incomplete_message_history() && !invalid_narrow;
}
```

can handle all the logic. 

Also, in each function where we have show_action getting repeated multiple times, we can extract it in a variable and calculate once.

shubham-padia commented on web/src/filter.ts:
```diff
@@ -1724,10 +1724,10 @@ export class Filter {
         );
     }
```
Last commit has a typo, oultine -> outline

shubham-padia reviewed (commented):
Thanks for the pass, much more readable now. left some comments.

pratik-pc committed:
inbox_ui: Read topic stream ID from its own menu element.

When updating the closed-compose reply button for an inbox topic row,
we found the topic's channel by walking from the row to its preceding
`.inbox-header` sibling and reading that header's `data-stream-id`.

The topic row's own `.inbox-topic-menu` element already carries
`data-stream-id`, so read it directly instead. This drops the fragile
dependence on the header being the immediately preceding sibling, with
no change in behavior.

pratik-pc committed:
compose_closed_ui: Fix recipient ids in DM reply button test.

The "DM policy disabled" test covers three recipient cases,
but the human-user case (Alice) overrode `pm_ids_string` with the
combined "31,33", and the following `human and bot` case had no
override of its own, so it silently re-ran that same input.
The single human-recipient case was therefore never exercised.

We give humer user case its own id ("31") and add the explicit "31,33"
override for the `human and bot` case, so each assertion tests the
recipients its comment describes.

pratik-pc commented:
@shubham-padia Addressed above reviews.

shubham-padia commented:
Looks good to me, merged, thanks @pratik-pc :)

Just a note for the future, leaving reviewer comments unresolved makes it easier for me to check that the feedback has been addressed or not.
